### ISSUE TITLE
LLVM assert abort while trying to take address of raw pointer in static

### ISSUE THREAD

**Issue by Kimundi** [rust-lang/rust#13973]
Testcase:
```rust
fn main() {
    unsafe {
        static BLOCK_UNSAFE_SAFE_PTR: &'static int = &*(0xdeadbeef as *int);
    }
}
```
Output:
```
rustc: /build/rust-git/src/rust/src/llvm/include/llvm/Support/Casting.h:237: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::GlobalVariable; Y = llvm::Value; typename llvm::cast_retty<X, Y*>::ret_type = llvm::GlobalVariable*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
fish: Job 1, “rustc bug_generic_static_function_ptr.rs ” terminated by signal SIGABRT (Abort)
```

**Comment by steveklabnik** [comment]
Triage:

```
rustc: /home/steve/src/rust/src/llvm/include/llvm/Support/Casting.h:237: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::GlobalVariable; Y = llvm::Value; typename llvm::cast_retty<X, Y*>::ret_type = llvm::GlobalVariable*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```

**Comment by cmr** [comment]
Confirm.

**Comment by JustAPerson** [comment]
**tl;dr**: *this bug still reproduces*
Updated for current Rust (on playpen as of today, 29 March 2015):

This *still fails the llvm assertion* that @steveklabnik mentioned.
```rust
fn main() {
    unsafe {
        static BLOCK_UNSAFE_SAFE_PTR: &'static isize = &*(0xdeadbeef as *const isize);
    }
}
```

**Comment by tamird** [comment]
I am no longer able to reproduce this on `rustc 1.0.0-dev (f46c4e158 2015-04-20) (built 2015-04-20)`

**Comment by jooert** [comment]
This still fails (when LLVM assertions are enabled) with
```
rustc: /path/to/rust/src/llvm/include/llvm/Support/Casting.h:237: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::GlobalVariable; Y = llvm::Value; typename llvm::cast_retty<X, Y*>::ret_type = llvm::GlobalVariable*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```

**Comment by pmarcelll** [comment]
This fails with a regular build error on the latest nightly:
```
test.rs:3:57: 3:86 error: raw pointers cannot be dereferenced in statics [E0396]
test.rs:3         static BLOCK_UNSAFE_SAFE_PTR: &'static isize = &*(0xdeadbeef as *const isize);
                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

**Comment by alexcrichton** [comment]
Agreed!

**Comment by cmr** [comment]
@alexcrichton is there a test for this? I didn't notice one.

**Comment by alexcrichton** [comment]
Probably not specifically, but PRs which often fix issues like this tend to have their own test cases anyway.
